### ISSUE TITLE
ResourcesManager global ResourceProvider configuration

### ISSUE THREAD

**Issue by Chad-Adams** [PSLCL/testing-framework#41]
Long term suggestion follows:

Add a new template step type that allows for a test script to declare name/value pairs (Provider Configuration) that is global to all bind types in that script (i.e. machine/network/person/other)

This would be generally optional, but might be required by some silo provider implementations.  Only one of these step types would be valid in any given test-script.  The dtf-runner implementation would need to guarantee that this step is processed before any bind steps.  It would simply call the implementations ResourcesManager.setAttributes(Map<String,String> attributes) method.

The individual ResourceProvider implementations controlled by the ResourcesManager can then be given access to the global/shared attributes.  Typical use will be a network provider and machine provider sharing commonly needed configuration.

Currently redundant/duplicate configuration setup in the test-script is required for the network bind and all machine binds.  This would remove that need for the shared configuration duplication.

**Comment by ghost** [comment]
Please provide a concrete use case. I do not like convoluting resource attributes with “global” attributes, and the whole idea of a global configuration goes against the isolated resource managers.

**Comment by Chad-Adams** [comment]
First, these would only be scoped to a given runId.  We have decided to only support "template level" reuse at this time (various reserve maps) and bind instance mapping is based on the runId of a given template node/tree (if nested).  These are not global to the ResourcesManager/Provider.

Also note that the only Apache Daemon dtf-runner init time startup configuration is to provide an installations desired DEFAULT "ResourceProvider Configuration" (i.e. ResourceDescription.getAttributes(), more specifically Bind Resource attributes).  We are trying to avoid any dtf-aws-resource installation specific configuration that is something other than these default bind attributes values that need injected if not specified by the test-script.  

Thus the first concrete example came up yesterday in that panicked flurry of emails I spammed everyone with - VPC/Subnet lifecyle handling.  

The machineBind's aws api createInstance call requires the vpcId and subnetId.  The networkBind also requires the vpcId and subnetId.  There is no guarantee of which bind will happen first (unless setId is required, which I first suggested, i.e. force networkBind to be first would be required by the ResourceProvider implementation).  With setId being rejected, the provider implementation is forced to implement the common attribute handle in two places instead of one.  And presents the test-script with the same redundancy problem.  With the current aws provider redundancy implementation there is a single config data object global to the two binds with both calling to initialize it.  A white-box savvy test-script writer could utilize setId on a network bind first and then would not have to duplicate in a following machine bind out of that setId.  Anyway I digress. (This feature request is really for the ResourcesManager/ResourceProvider implementors)

The next one is a short prefix name that I allow the test script to optionally set.  The default value is "dtf".  I then do some fairly rich tagging and naming of all AWS objects created with a "prefix-midstr-post" scheme that allows for quick visual (aws console) as well programmatic failure cleanups etc., thus dtf-ex2-part-of-templateId, dtf-vpc-part-of-templateId etc.  With aws object creation in machine and network binds (unknown at this point if person binds will utilize any), the same redundancy issue exists for this attribute.

The last one I've come across so far are timeout values.  I currently use two flavors of these (per bind type that uses aws api calls).  Some calls can take several minutes before completion, some are relatively fast.  I currently default one to about 1 minute and one to about 15 minutes (x3, machine, network, person).

Basically the questions are:
1. would it be ok for a ResourceProvider implementation to require a test-script to use a setId to remove their need for known race condition handling?
2. More formally would it be ok for a ResourcesManager implementation to require a test-script to declare a test run scoped global "setId" as it were, required to execute before any binds, for a subset of well defined attributes if those attributes are optionally used in the script.

hope that makes sense.

**Comment by ghost** [comment]
The problem is that anything that appears in the script changes the hash, and thus produces a “different” test template. Anything that doesn’t relate directly and consequentially to the test itself (things like the operating system choice, versioning, etc.) is not appropriate for the template itself.

If I understand your emails from yesterday, things like the AWS configuration seem to apply to the resource manager instance, not the test instance. For example, I could imagine two separate instances of an AWS resource manager that used different vpcId/subnetId combinations in order to avoid conflicts in the instances that they created. It would not be appropriate for the test itself to specify anything about that configuration information because the test could run equally well using either resource provider (meaning either combination of vpcId/subnetId).

Having said that, it is perfectly OK for the test to have access to that information, through the attribute map, at run time. That is OK because the run time attribute map is not part of the test definition (not part of the hash). That is the mechanism for run-time information to make it out to the running test (for example, to get an IP address), but doesn’t work for pre-run configuration information.

I hope that makes sense. It is a critical difference because it goes to the entire architecture/design of the test definitions. Basically, anything that affects the hashing (artifacts and templates) has a  huge impact on the system.

**Comment by Chad-Adams** [comment]
I personally do not have any exposure to all of the templateId (hash) concepts.  I’ve only had involvement in the leaf node of the whole system.  I assume things are on track as everything has been passed by Joe/Clint.  However statements like:

“Anything that doesn’t relate directly and consequentially to the test itself (things like the operating system choice, versioning, etc.) is not appropriate for the template itself”

do concern me, as it makes no sense from my API vantage point.  This is an unusual project, in that I think we all have had a hard time with terminology etc.  However, I believe the “magic” of all of this is solid above me and by the time it hits my api entry points, all will be well.

I was assigned to the ResourceProvider aws implementation.  When I started there was only the concept of three types, Machine, Network, Person.  To keep things simple here let’s say each of these have the single entry point of:

public Future<? extends ResourceInstance> bind(ReservedResource resource) throws ResourceNotReservedException;

Coming into my subsystem, all I know about is an implementation of the ResourceDescription interface which provides nothing more than the requested resource’s coordinates (templateId, resourceId, runId) and the attributes map (“Provider Configuration”).  I only use the coordinates for tagging and map keys.

In my mind, by whatever magic, ultimately the test-script is the master source of all Provider Configuration.  Default values for all optional Provider Configuration attributes are supported (via installation time properties file)  But ultimately it’s the test engineer script writer who declares the scripts defaults for the installation and then overrides the defaults as they seem fit.

Thus the ResourceProvider implementation level has no concept of hashing, testId’s, runId’s etc.  These are all resolved in the hierarchy above it. 

I introduced the ResourcesManager into the ResourceProvider subsystem.  For the aws implementation when I realized early on that I needed a structure above the three independent implementations.  And yet there was no singular entry point from the dtf-runner for this.  The ResourcesManager became that startup entry point.  The dtf-runner code now utilizes some release resources methods on this new manager which are based on the temple level (release all resources, machine, bind, person for this given templateId ), allowing it to more easily handle reuse at the template level.

When it came down to the vpc/subnet implementation (and/or secure group under that) the coordinates of interest to the network now become:

TemplateId, wild-card, runId

Where several ResourceIds are associated with the unique TemplateId/runId pairs.  The ResourcesManager of course contains the maps for these as the ResourceIds could be multiples of machine, network and person.  The non deterministic nature of the bind call order of these \* ResourceId’s to the given TemplateId/RunId pairs is the problem that arose and is what I’m proposing this change for.

**Comment by ghost** [comment]
And this is the conflict:

“Anything that doesn’t relate directly and consequentially to the test itself (things like the operating system choice, versioning, etc.) is not appropriate for the template itself”

vs.

“In my mind, by whatever magic, ultimately the test-script is the master source of all Provider Configuration.”

Earlier in your message you use a resource coordinate (templateId, resourceId, runId). That is a very different thing than “test-script”. I agree that the coordinate you mention is the master source of configuration, just not that the test-script/templateId is the correct place to put it. As I stated in a previous message, the problem isn’t using the attribute map - at the point you are executing the attribute map may have been extended through the resource manager (the resourceId and runId axis of your resource coordinate). Thus, your API may be fine, only the original statement about where the information comes from originally that causes issues.

If there is still confusion, we should schedule a meeting to clarify the technical specification.

**Comment by jwclark** [comment]
Not enough benefit for changes required.
